[profiling] Avoid crashing when pcntl_fork is called

morrisonlevi opened the pull request:
### Description

The profiler doesn't support forking yet. However, it shouldn't crash.
This PR reduces some crash surface, but there is more to be done.

Since the tracer already has an API for overwriting the internal
function handler of a given function, I moved those bits into their own
file and made a note that the profiler also compiles these files.

I investigated adding full forking support but decided it would be too
much work to do right now. However, I left some comments about what I
learned, and did leave in some refactoring.

### Readiness checklist
- [x] Changelog has been added to the release document.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [x] Appropriate labels assigned.
- [x] Milestone is set.

bwoebi commented on profiling/src/pcntl.rs:
```diff
@@ -0,0 +1,118 @@
+        Ok(pid) => {
+            // The child gets pid of 0. For now, stop profiling for safety.
+            if pid == 0 {
+                match PROFILER.lock() {
```
May that not ... just block indefinitely when the lock is held by the other thread at time of forking? At least docs aren't mentioning mutex poisoning in context of forking.

morrisonlevi replied:
Yes, that can happen in theory. In practice, I don't support ZTS and the profiler is only used on the PHP thread (caveat, it's also called by the tracer when a trace finishes, so please don't do that in from a non-PHP thread 😅 ).

But, that doesn't mean I'm in the clear. I'm not sure about all of the internals of the objects that the Profiler stores, for instance.

bwoebi replied:
Why does it need a mutex at all, when it's only used on a single thread? Or is this in anticipation of future ZTS support?

morrisonlevi replied:
1. Yes, it's in anticipation of ZTS support. Even if it's not a critical part of the business, I think it's smart to have a plan there.
2. Having a global mutable variable like this requires safe abstractions in Rust or using unsafe code. Given the above point, it made sense to me (at the time anyway) to just go ahead and use the mutex.

bwoebi commented on profiling/src/pcntl.rs:
```diff
@@ -0,0 +1,118 @@
+
+/* Safety: CStr invariants are met (at least one byte, last byte is null,
+ * no null bytes before that). Note that they require their own unsafe
+ * blocks because they are const, despite being in th
```
in th?

morrisonlevi replied:
Removed. I used to define these inside of `startup` since they are only used there, but I had to add `unsafe` around them even though `startup` was already an `unsafe` function, so I was starting to document that.

morrisonlevi commented:
I did a bit more auditing. The biggest problems with multi-threaded code and forking are locks, for example:

- Every 10ms a thread will acquire a lock which contains a list of addresses of the VM interrupt globals, and then iterate over the list to trigger them all. The lock is also acquired at the start and end of every request, to add and remove the VM interrupt address respectively. For non-ZTS we could have a simpler synchronization mechanism, but since I wanted to make sure our foundations were ZTS compatible, that's how it is today.
- Printing anything to stderr will have a lock, such as log messages.
- I haven't fully unwrapped the crossbeam channels abstractions, so there could be locking in there too. So far I'm only seeing atomics, though.

If locks are held at the time of a fork and it's possible for the thread which forked to also get that lock, then we're likely to get a deadlock eventually.

So... the current state of the PR is that several opportunities for _crashing_ have been fixed, but there are still other scenarios that need more work. I'm thinking we should merge this as it's a clear improvement, but it's insufficient, so I should continue working on addressing these issues.

bwoebi commented:
I agree that we shall merge the PR as is - at least to prevent the crash. Iterative progress is much better than having no progress at all here :-)

bwoebi reviewed (approved):
Looks good :-)

r1viollet commented:
We had some discussions about implications of "forgetting" the old instance of the old profiler. It will increase memory overhead of a single fork. For sure this is an improvement over crashes. The challenges behind calling the actual destructors and allowing a new profiler to start are not so easy, so I agree with this first step ! I'll review some more.

r1viollet reviewed (approved):
LGTM
